[CAS] LLVMCAS implementation

cachemeifyoucan opened the pull request:
Adds Content Addressable Storage implementation for LLVM, which includes:
* InMemoryCAS (non-persistent)
* OnDiskCAS (persistent)
Also comes with various level of abstraction for create/query/chaining CAS instance.

cachemeifyoucan committed:
[ADT] Add TrieRawHashMap

Implement TrieRawHashMap which stores objects into a Trie based on the
hash of the object.

User needs to supply the hashing function and guarantees the uniqueness of
the hash for the objects to be inserted. Hash collision is not
supported

cachemeifyoucan committed:
[CAS] Add LLVMCAS library with InMemoryCAS implementation

Add llvm::cas::ObjectStore abstraction and InMemoryCAS as a in-memory
CAS object store implementation.

The ObjectStore models its objects as:
* Content: An array of bytes for the data to be stored.
* Refs: An array of references to other objects in the ObjectStore.
And each CAS Object can be idenfied with an unqine ID/Hash.

ObjectStore supports following general action:
* Expected<ID> store(Content, ArrayRef<Ref>)
* Expected<Ref> get(ID)

It also introduces following types to interact with a CAS ObjectStore:
* CASID: Hash representation for an CAS Objects with its context to help
  print/compare CASIDs.
* ObjectRef: A light-weight ref for an object in the ObjectStore. It is
  implementation defined so it can be optimized for
  read/store/references depending on the implementation.
* ObjectHandle: A CAS internal light-weight handle to an loaded object in the
  ObjectStore. Underlying data for the object is guaranteed to be
  available and no error handling is required to access data. This is
  not exposed to the users of CAS from ObjectStore APIs.
* ObjectProxy: A proxy for the users of CAS to interact with the data
  inside CAS Object. It bundles a ObjectHandle and an ObjectStore
  instance.

Differential Revision: https://reviews.llvm.org/D133716

cachemeifyoucan committed:
[FileSystem] Allow exclusive file lock

Add parameter to file lock API to allow exclusive file lock. Both Unix
and Windows support lock the file exclusively for write for one process
and LLVM OnDiskCAS uses exclusive file lock to coordinate CAS creation.

cachemeifyoucan committed:
[CAS] Add implementation for current OnDisk CAS + abstractions

Add current downstream cas API and implementation that includes
OnDiskCAS implementation, different level of abstractions for CAS,
different utilities.

cachemeifyoucan committed:
[UnitTest] Add unit-test for LLVMCAS

cachemeifyoucan committed:
[CAS] Add llvm-cas test

dwblaikie commented:
(be great if this could be broken up a bit more - like the TrieRawHashMap looks like it'd be a fine standalone patch to introduce the data structure and its unit tests - maybe similarly with the other VirtualOutputBackends patch, the raw_ostream_proxy might be separable (& it'd be nice if the other parts of the functionality were built up more incrementally too))

But maybe it's just not practical - it's unfortunate when things get to this state, though I understand the value of prototyping out of tree - it makes it hard to adequately review things when they come into the project this way.

cachemeifyoucan commented:
> (be great if this could be broken up a bit more - like the TrieRawHashMap looks like it'd be a fine standalone patch to introduce the data structure and its unit tests - maybe similarly with the other VirtualOutputBackends patch, the raw_ostream_proxy might be separable (& it'd be nice if the other parts of the functionality were built up more incrementally too))
> 
> But maybe it's just not practical - it's unfortunate when things get to this state, though I understand the value of prototyping out of tree - it makes it hard to adequately review things when they come into the project this way.

Both of the things you mentioned above are separate commits in those PRs already and I will continue to break apart more functions if needed. I am happy to put up individual PRs if that is better for reviewers while shrinking this mega PR.

dwblaikie commented:
> > (be great if this could be broken up a bit more - like the TrieRawHashMap looks like it'd be a fine standalone patch to introduce the data structure and its unit tests - maybe similarly with the other VirtualOutputBackends patch, the raw_ostream_proxy might be separable (& it'd be nice if the other parts of the functionality were built up more incrementally too))
> > But maybe it's just not practical - it's unfortunate when things get to this state, though I understand the value of prototyping out of tree - it makes it hard to adequately review things when they come into the project this way.
> 
> Both of the things you mentioned above are separate commits in those PRs already and I will continue to break apart more functions if needed. I am happy to put up individual PRs if that is better for reviewers while shrinking this mega PR.

Ah, fair enough. Not ideal (it's still sort of one review for the whole set of changes & any updates get disjoint (because they've got to be stacked on top of the already complex stack, rather than added into the stack back where the original change was made, etc)

Such is life - eventually we'll get some better dependent-patch review process I hope.

cachemeifyoucan commented on llvm/lib/Support/TrieRawHashMap.cpp:
```diff
@@ -0,0 +1,493 @@
+    OS << char(Digit + '0');
+  else
+    OS << char(Digit - 10 + 'a');
+}
```
The complication here is that the prefix can have N bits, where N is not a multiple of 4 or 8. It was printing as much hex as possible, then print the remaining in `[]` as binary. I change it to print as much complete bytes as possible, so we can use the existing hex printing function.

cachemeifyoucan committed:
[CAS] Add support for PluginCAS

Allow loading external CAS implementation via PluginCAS. In this patch,
it adds:
* C APIs that can be implemented by plugin, from which LLVM can load the
  dylib to use external CAS implementation
* A PluginCAS, that implements vending external CAS implementation as
  llvm ObjectStore and ActionCache class.
* A libCASPluginTest dylib, that provides example external CAS
  implementation that wraps LLVM CAS for testing purpose.
* Add a unified way to load external CAS implementation.

cachemeifyoucan committed:
Add modulemap for LLVM CAS

cachemeifyoucan committed:
Try fix msvc test failures

aganea commented:
Hello @cachemeifyoucan! Do you still plan on working on these PRs? There were some unaddressed comments/requests for changes left last year.

+1 with @dwblaikie - this would need to be split in smaller patches, there's just too much code to review. Can you send a PR for the `Path` changes? Can you send another PR for the basic/most indivisible interfaces with mock implementations in unit tests? I think you could then send specific implementations one by one. And finish with `llvm-cas` and the LLVM C APIs? My impression is that everyone agrees that we need all this.

aganea commented on llvm/include/llvm/CAS/CASReference.h:
```diff
@@ -0,0 +1,207 @@
+/// ObjectHandle encapulates a *loaded* object in the CAS. You need one
+/// of these to inspect the content of an object: to look at its stored
+/// data and references.
+class ObjectHandle : public ReferenceBase {
```
I just skimmed over the PR, but it still unclear why `ObjectHandle` is needed. Is this just for semantic purposes? Why can't we just have `ObjectRef`?

cachemeifyoucan replied:
The current intention for the `ObjectHandle` class is that it is only exposed to the CAS implementation. There is a documentation file https://github.com/llvm/llvm-project/blob/281d1781dc4cb9edd457600b5133afcf25bda544/llvm/docs/ContentAddressableStorage.md that explains how to use the types (together with doxyen comments).

aganea commented:
FWIW we build something like this a while ago: https://github.com/legion-labs/legion/tree/main/crates/lgn-content-store

cachemeifyoucan commented:
> Hello @cachemeifyoucan! Do you still plan on working on these PRs? There were some unaddressed comments/requests for changes left last year.
> 
> +1 with @dwblaikie - this would need to be split in smaller patches, there's just too much code to review. Can you send a PR for the `Path` changes? Can you send another PR for the basic/most indivisible interfaces with mock implementations in unit tests? I think you could then send specific implementations one by one. And finish with `llvm-cas` and the LLVM C APIs? My impression is that everyone agrees that we need all this.

Hi @aganea Thanks for helping with the review! I was distracted with other obligation that I temporary unable to follow up with the reviews and I definitely intended to pick this up ASAP. Note this PR was created just as we transition to GitHub PR that there isn't a clear workflow for how to handle a big PR like this. I was breaking up the big commit into smaller reviews (like https://github.com/llvm/llvm-project/pull/69528) while using this PR to provide a larger context. I might take some time to rewrite this PR using `spr`. I will remember to add you as reviewer in future PR.


> FWIW we build something like this a while ago: https://github.com/legion-labs/legion/tree/main/crates/lgn-content-store

Cool! Glad to see more community members are doing something similar and we hope we can work together!

Naville commented:
Is there any progress on this?
We'd **love** to use CAS across our build infrastructure

cachemeifyoucan committed:
Rebase change to latest

cachemeifyoucan committed:
format fixup

cachemeifyoucan commented:
Most content in this PR has been split into following stacked PR:
* [CAS] Add LLVMCAS library with InMemoryCAS implementation https://github.com/llvm/llvm-project/pull/114096
* [CAS] Add ActionCache to LLVMCAS Library https://github.com/llvm/llvm-project/pull/114097
* [FileSystem] Allow exclusive file lock https://github.com/llvm/llvm-project/pull/114098
* [CAS] Add MappedFileRegionBumpPtr https://github.com/llvm/llvm-project/pull/114099
* [CAS] Add OnDiskHashMappedTrie https://github.com/llvm/llvm-project/pull/114100
* [CAS] Add OnDiskGraphDB and OnDiskKeyValueDB https://github.com/llvm/llvm-project/pull/114102
* [CAS] Add OnDiskCAS https://github.com/llvm/llvm-project/pull/114103
* [CAS] Add llvm-cas tools to inspect on-disk LLVMCAS https://github.com/llvm/llvm-project/pull/114104

Going to close this PR once everything is moved. Please head to the new PRs for easier review. Thanks.
